File: Microservice_training/productservice/src/main/java/com/wisdom/services/ProductService.java
package com.wisdom.services;

import com.wisdom.dto.request.ProductDTO;
import com.wisdom.dto.respone.ProductResponse;
import com.wisdom.entity.Product;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ProductService {
    Product createProduct(ProductDTO productDTO);

    ProductResponse todo(Product product);

    List<Product> getAll();

    Product getProductById(int productId);

    Product updateProduct (int productId, ProductDTO productDTO);

    void deleteProduct(int productId);

    Page<Product> getAllByCategory(int category_id, Pageable pageable); // (with pagination)

    Product updateProductCategory(int productId, int newCategoryId);

    Page<Product> getPaginatedProducts(Pageable pageable);
}

File: fe_page_admin/src/services/productService.js
import ApiService from './api';

export const productService = {
  // Get all products with pagination
  getProducts: (page = 0, size = 10) => 
    ApiService.get(`/products?page=${page}&size=${size}`),
  
  // Get product by ID
  getProductById: (id) => ApiService.get(`/products/${id}`),
  
  // Create new product
  createProduct: (productData) => ApiService.post('/products', productData),
  
  // Update product
  updateProduct: (id, productData) => ApiService.put(`/products/${id}`, productData),
  
  // Delete product
  deleteProduct: (id) => ApiService.delete(`/products/${id}`),
  
  // Update product category
  updateProductCategory: (productId, categoryId) => 
    ApiService.request(`/products/${productId}/category/${categoryId}`, { method: 'PATCH' }),
};
